staging/src/k8s.io/apiserver/pkg/util/flowcontrol/rule.go

```diff
@@ -0,0 +1,146 @@
+		}
+		return serviceaccount.MatchesUsername(subject.ServiceAccount.Namespace, subject.ServiceAccount.Name, user.GetName())
+	default:
+		return false
+	}
+}
+
+// serviceAccountMatchesNamespace checks whether the provided service account username matches the namespace, without
+// allocating. Use this when checking a service account namespace against a known string.
+// This is copied from `k8s.io/apiserver/pkg/authentication/serviceaccount::MatchesUsername` and simplified to not check the name part.
+func serviceAccountMatchesNamespace(namespace string, username string) bool {
+	const (
+		ServiceAccountUsernamePrefix    = "system:serviceaccount:"
+		ServiceAccountUsernameSeparator = ":"
+	)
+	if !strings.HasPrefix(username, ServiceAccountUsernamePrefix) {
+		return false
+	}
+	username = username[len(ServiceAccountUsernamePrefix):]
+
+	if !strings.HasPrefix(username, namespace) {
+		return false
+	}
+	username = username[len(namespace):]
+
+	return strings.HasPrefix(username, ServiceAccountUsernameSeparator)
+}
+
+func matchesResourcePolicyRule(digest RequestDigest, policyRule rmtypesv1alpha1.ResourcePolicyRule) bool {
+	if !matchPolicyRuleVerb(policyRule.Verbs, digest.RequestInfo.Verb) {
+		return false
+	}
+	if !digest.RequestInfo.IsResourceRequest {
+		return false
+	}
+	if !matchPolicyRuleResource(policyRule.Resources, digest.RequestInfo.Resource) {
+		return false
+	}
+	if !matchPolicyRuleAPIGroup(policyRule.APIGroups, digest.RequestInfo.APIGroup) {
+		return false
+	}
+	if len(digest.RequestInfo.Namespace) == 0 {
+		return policyRule.ClusterScope
+	}
+	return containsString(digest.RequestInfo.Namespace, policyRule.Namespaces, rmtypesv1alpha1.NamespaceEvery)
+}
+
+func matchesNonResourcePolicyRule(digest RequestDigest, policyRule rmtypesv1alpha1.NonResourcePolicyRule) bool {
+	if !matchPolicyRuleVerb(policyRule.Verbs, digest.RequestInfo.Verb) {
+		return false
+	}
+	if digest.RequestInfo.IsResourceRequest {
+		return false
+	}
+	return matchPolicyRuleNonResourceURL(policyRule.NonResourceURLs, digest.RequestInfo.Path)
+}
+
+func matchPolicyRuleVerb(policyRuleVerbs []string, requestVerb string) bool {
+	return containsString(requestVerb, policyRuleVerbs, rmtypesv1alpha1.VerbAll)
+}
+
+func matchPolicyRuleNonResourceURL(policyRuleRequestURLs []string, requestPath string) bool {
+	return containsString(requestPath, policyRuleRequestURLs, rmtypesv1alpha1.NonResourceAll)
+}
+
+func matchPolicyRuleAPIGroup(policyRuleAPIGroups []string, requestAPIGroup string) bool {
+	return containsString(requestAPIGroup, policyRuleAPIGroups, rmtypesv1alpha1.APIGroupAll)
+}
+
+func matchPolicyRuleResource(policyRuleRequestResources []string, requestResource string) bool {
+	return containsString(requestResource, policyRuleRequestResources, rmtypesv1alpha1.ResourceAll)
+}
+
+func containsString(x string, list []string, wildcard string) bool {
+	if len(list) == 1 && list[0] == wildcard {
+		return true
+	}
+	for _, y := range list {
```

Comment:
For sufficiently large lists it will be better to pre-construct a sets.String. (not asking for a change in this PR)

Reply:
Yeah, I might be inclined to leave that until we see long lists.